clippy_lints/src/pub_underscore_fields.rs

```diff
@@ -0,0 +1,83 @@
+use clippy_utils::attrs::is_doc_hidden;
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::is_path_lang_item;
+// use rustc_ast::ast::{Item, ItemKind, VisibilityKind};
+use rustc_hir::{FieldDef, Item, ItemKind, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::Visibility;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks whether any field of the struct is prefixed with an `_` (underscore) and also marked
+    /// `pub` (public)
+    ///
+    /// ### Why is this bad?
+    /// Fields prefixed with an `_` are inferred as unused, which suggests it should not be marked
+    /// as `pub`, because marking it as `pub` infers it will be used.
+    ///
+    /// ### Example
+    /// ```rust
+    /// struct FileHandle {
+    ///     pub _descriptor: usize,
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// struct FileHandle {
+    ///     _descriptor: usize,
+    /// }
+    /// ```
+    ///
+    /// // OR
+    ///
+    /// ```
+    /// struct FileHandle {
+    ///     pub descriptor: usize,
+    /// }
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub PUB_UNDERSCORE_FIELDS,
+    pedantic,
+    "struct field prefixed with underscore and marked public"
+}
+declare_lint_pass!(PubUnderscoreFields => [PUB_UNDERSCORE_FIELDS]);
+
+impl<'tcx> LateLintPass<'tcx> for PubUnderscoreFields {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        // This lint only pertains to structs.
+        let ItemKind::Struct(var, _) = &item.kind else {
+            return;
+        };
+
+        let is_visible = |field: &FieldDef<'_>| {
+            let parent = cx.tcx.parent_module_from_def_id(field.def_id);
+            let grandparent = cx.tcx.parent_module_from_def_id(parent.into());
+            let visibility = cx.tcx.visibility(field.def_id);
+
+            let case_1 = parent == grandparent && !field.vis_span.is_empty();
+            let case_2 = visibility != Visibility::Restricted(parent.to_def_id());
+
+            case_1 || case_2
+        };
```

Comment:
Can try `cx.effective_visibilities.is_exported(field.def_id))` for this

Reply:
I tried that, unfortunately it didn't work :/

Reply:
I think the issue is the test file, none of the fields in it are actually publicly reachable. You'd have to move the definitions outside of `fn main() {}` and make the `inner`/`inner2` modules public

Reply:
`is_exported` doesn't lint for `pub(...)`, and it seems like it has some problems with fields of type `Option<T>` for some reason?

I don't see it as a requirement.

Reply:
The actual visibility is what the lint is about though

> Fields prefixed with an `_` are inferred as unused, which suggests it should not be marked as `pub`, because marking it as `pub` infers it will be used.

`pub(crate)`, or `pub` where the field is not publicly reachable does not mark the field as used

What's the issue with `Option` fields?

Reply:
Hmm, now I'm not sure.
We could lint for any field that is public and its name starts with `_`, which is what the name of lint implies. Or we could analyze the description a little bit more and just lint when the `pub` is actually useful.